sklearn/inspection/_h_statistic.py

```diff
@@ -0,0 +1,317 @@
+"""Friedman and Popescu's H-Statistic"""
+
+import itertools
+
+import numpy as np
+from scipy import sparse
+
+from ..base import is_classifier, is_regressor
+from ..utils import Bunch, check_array
+from ..utils._indexing import _get_column_indices, _safe_assign, _safe_indexing
+from ..utils._param_validation import (
+    HasMethods,
+    Integral,
+    Interval,
+    Real,
+    validate_params,
+)
+from ..utils.random import sample_without_replacement
+from ..utils.validation import _check_sample_weight, check_is_fitted
+
+
+def _calculate_pd_brute_fast(
+    pred_fun, X, feature_indices, grid, sample_weight=None, reduce_binary=False
+):
+    """Fast version of _calculate_partial_dependence_brute()
+
+    Returns np.array of size (n_grid, output_dim).
+    """
+
+    # X is stacked n_grid times, and grid columns are replaced by replicated grid
+    n = X.shape[0]
+    n_grid = grid.shape[0]
+
+    X_stacked = _safe_indexing(X, np.tile(np.arange(n), n_grid), axis=0)
+    grid_stacked = _safe_indexing(grid, np.repeat(np.arange(n_grid), n), axis=0)
+
+    if hasattr(X, "iloc"):  # pandas<2 does not allow "values" to have repeated indices
+        grid_stacked = grid_stacked.reset_index(drop=True)
```

Comment:
I have the feeling that those lines belong to `_safe_indexing`. The point of `_safe_indexing` is precisely to hide the pandas-specific things.

It might be worth opening in a side PR with a dedicated non-regression test for #28931.

Reply:
Feel free to keep temporarily keep those lines here but with a TODO comment that links to https://github.com/scikit-learn/scikit-learn/issues/28931 and/or the new side PR.

Reply:
Okay! Do you have a good idea how to get rid of the still quite hacky 

```py
    try:
        ax = 0 if grid.shape[1] > 1 else None  # np.unique works better in 1 dim
        _, ix, ix_reconstruct = np.unique(
            grid, return_index=True, return_inverse=True, axis=ax
        )
        grid = _safe_indexing(grid, ix, axis=0)
        compressed_grid = True
    except (TypeError, np.AxisError):
        compressed_grid = False

    pd_values = _calculate_pd_brute_fast(
        pred_fun,
        X=X,
        feature_indices=feature_indices,
        grid=grid,
        sample_weight=sample_weight,
        reduce_binary=reduce_binary,
    )

    if compressed_grid:
        pd_values = pd_values[ix_reconstruct]
```

The grid equals one or two columns from the data. The snipped tries to remove duplicated rows (to save a lot of time in calculating partial dependence for discrete features). The last step merges the result back to the original row order.

I see these alternatives:

1. Use joblib memoization on `_calculate_pd_brute_fast()` instead of grid deduplication.
2. Write a neat `_safe_unique()` function that would return the unique rows along with the reconstruction information:

```py
from sklearn.utils._indexing import _safe_indexing

def _safe_unique(X):
    axis=None
    if isinstance(X, np.ndarray):
        groups = X.copy()
        if len(X.shape) > 1 and X.shape[1] > 1:
            axis = 0
        
    if hasattr(X, "iloc"):
        groups = X.groupby(X.columns.to_list(), sort=False, dropna=False).ngroup()

    _, ix, ix_reconstruct = np.unique(
        groups, return_index=True, return_inverse=True, axis=axis
    )
    return(_safe_indexing(X, indices=ix, axis=0), ix_reconstruct)

df_unique, reverse_ix = _safe_unique(df[["b"]].to_numpy())
df_unique
```